tools/pd-ctl/pdctl/command/patrol_command.go

```diff
@@ -0,0 +1,389 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package command
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"slices"
+	"time"
+
+	"github.com/docker/go-units"
+	"github.com/spf13/cobra"
+	"google.golang.org/grpc"
+
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb/pkg/util/codec"
+
+	pd "github.com/tikv/pd/client"
+	"github.com/tikv/pd/client/clients/router"
+	"github.com/tikv/pd/client/http"
+	"github.com/tikv/pd/client/opt"
+	"github.com/tikv/pd/client/pkg/caller"
+	"github.com/tikv/pd/pkg/utils/typeutil"
+)
+
+const (
+	msgSize                = 16 * units.MiB
+	defaultLimit           = 8192
+	statusMergeFailed      = "merge_failed"
+	statusMergeSkipped     = "merge_skipped"
+	statusMergeRequestSent = "merge_request_sent"
+	maxMergeRetries        = 10
+	mergeRetryDelay        = 6 * time.Second
+)
+
+// PatrolResult defines the structure for JSON output of each processed region.
+type PatrolResult struct {
+	RegionID    uint64 `json:"region_id"`
+	Key         string `json:"key"`
+	TableID     int64  `json:"table_id"`
+	Status      string `json:"status"`
+	Description string `json:"description,omitempty"`
+}
+
+// PatrolResults defines the structure for the final JSON output of the patrol command.
+type PatrolResults struct {
+	ScanCount    int               `json:"scan_count"`
+	ScanDuration typeutil.Duration `json:"scan_duration"`
+	Count        int               `json:"count"`
+	Results      []PatrolResult    `json:"results"`
+}
+
+// NewPatrolCommand creates the patrol subcommand.
+func NewPatrolCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:               "patrol",
+		Short:             "Patrol regions to find special keys and optionally merge them. Note that this command is temporary for tiflash#10147.",
+		PersistentPreRunE: requirePDClient,
+		Run:               patrolCommandFunc,
+	}
+	cmd.Flags().Int("limit", defaultLimit, "Limit of regions to scan per batch. If limit is not positive, it means no limit.")
+	cmd.Flags().Bool("enable-auto-merge", false, "Enable automatic region merge for regions with special keys.")
+	return cmd
+}
+
+func patrolCommandFunc(cmd *cobra.Command, _ []string) {
+	ctx, cancel := context.WithCancel(cmd.Context())
+	defer cancel()
+
+	// Create a pd client
+	opts, err := getSecurityOpt(cmd)
+	if err != nil {
+		cmd.Printf("Failed to parse TLS options: %v\n", err)
+		return
+	}
+	addr := getEndpoints(cmd)
+	cli, err := pd.NewClientWithContext(ctx, caller.PDCtlComponent, addr, opts,
```

Comment:
Does that mean pd-ctl will use both HTTP and gRPC to access the server right now?

Reply:
It only uses gRPC when call patrol sub command.

Reply:
Is it necessary?